Add containerd-snapshotter documentation (CVM-1914) (#164)
cvmfs/doc-cvmfs: cpt-containers.rst

The default socket is ``/run/containerd-cvmfs-grpc/containerd-cvmfs-grpc.sock``.
This socket is created automatically by the snapshotter when building the binary, if it does not exist.

To build the binary, use the following commands:

::

    cd <source directory>
    make

A new ``/out`` folder is created with the binary ``cvmfs-snapshotter``.
It is necessary to configure containerd to use this new remote snapshotter.
A basic configuration file would look like:

```
# tell containerd to use this particular snapshotter
[plugins."io.containerd.grpc.v1.cri".containerd]
  snapshotter = "cvmfs-snapshotter"
  disable_snapshot_annotations = false

# tell containerd how to communicate with this snapshotter
[proxy_plugins]
  [proxy_plugins.cvmfs-snapshotter]
    type = "snapshot"
    address = "/run/containerd-cvmfs-grpc/containerd-cvmfs-grpc.sock"
```
and it should be stored at ``containerd-remote-snapshotter/script/config/etc/containerd-cvmfs-grpc``.
One can directly run the binary indicating the path to the configuration file by using the ``--config`` option.

How to use the CernVM-FS Snapshotter
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The binary accepts different command line options:

> * ``--log-level`` - address for the snapshotter's GRPC server. 

Reviewer: This seems somewhat off

The default one is ``/run/containerd-cvmfs-grpc/containerd-cvmfs-grpc.sock``
* ``--config`` - path to the configuration file.
* ``--address`` - logging level [trace, debug, info, warn, error, fatal, panic].
* ``--root`` - path to the root directory for this snapshotter.
The default one is ``/var/lib/containerd-cvmfs-grpc``.

> By default, the repository used to search for the layers is ``unpacked.cern.ch``.

Reviewer: It would be helpful to show a sample `config.toml` file

the CernVM-FS container integration.

``containerd`` remote-snapshotter plugin
----------------------------------------

CernVM-FS integration with ``containerd`` is achieved by the snapshotter plugin,
a specialized component responsible for assembling all the layers of container
images into a stacked filesystem that ``containerd`` can use.
The snapshotter takes as input the list of required layers and outputs a directory
containing the final filesystem. It is also responsible to clean-up the output
directory when containers using it are stopped.

From version 1.4.0, containerd introduced the concept of remote snapshotter.

> It allows starting containers in which the filesystem is provided externally from the containerd machinery.

Reviewer: ```suggestion
It allows starting containers in which the container root filesystem is provided by an external plugin.
```

> Therefore, there is no need to download all the layers for each image, getting rid of the pulling time.

Reviewer: ```suggestion
Given such a suitable plugin (remote snapshotter), there is no need to download all the layers for each image, which can greatly improve the start-up time of containers.
```

> Overall, this new mechanism brings down the time to start-up a new container image.

Reviewer: ```suggestion
The CernVM-FS remote snapshotter uses this new capability to mount the container image layers directly from a CernVM-FS repository.
```

> We exploit this new capability to mount OCI layers directly from a filesystem on the local machine.

Reviewer: Let's remote this paragraph

Reviewer: Please remove this paragraph.

We focus on layers provided by CernVM-FS, but with minor changes is possible to mount layers from any
filesystem, like NFS. If the layers are not in the local filesystem, `containerd` simply follow the
standard path downloading them from the standard docker registry.

Configuration
~~~~~~~~~~~~~

> This remote snapshotter communicates with ``containerd`` via gRPC over linux socket.

Reviewer: ```suggestion
The CernVM-FS remote snapshotter communicates with ``containerd`` via gRPC over a UNIX domain socket.
```

The default socket is ``/run/containerd-cvmfs-grpc/containerd-cvmfs-grpc.sock``.
This socket is created automatically by the snapshotter when building the binary, if it does not exist.

To build the binary, use the following commands:

::

    cd <source directory>
    make

> Then, a new ``/out`` folder is created with the binary ``cvmfs-snapshotter``.

Reviewer: ```suggestion
A new ``/out`` folder is created with the binary ``cvmfs-snapshotter``.
```

It is necessary to configure containerd to use this new remote snapshotter.
A basic configuration file would look like:

```
# tell containerd to use this particular snapshotter
[plugins."io.containerd.grpc.v1.cri".containerd]
  snapshotter = "cvmfs-snapshotter"
  disable_snapshot_annotations = false

# tell containerd how to communicate with this snapshotter
[proxy_plugins]
  [proxy_plugins.cvmfs-snapshotter]
    type = "snapshot"
    address = "/run/containerd-cvmfs-grpc/containerd-cvmfs-grpc.sock"
```
and it should be stored at ``containerd-remote-snapshotter/script/config/etc/containerd-cvmfs-grpc``.

Testing
~~~~~~~

This plugin is tested using ``kind``.

```
$ docker build -t cvmfs-kind-node https://github.com/cvmfs/containerd-remote-snapshotter.git
$ cat kind-mount-cvmfs.yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
  extraMounts:
    - hostPath: /cvmfs/unpacked.cern.ch
      containerPath: /cvmfs/unpacked.cern.ch

$ kind create cluster --config kind-mount-cvmfs.yaml --image cvmfs-kind-node
```
At this point, it is possible to use ``kubectl`` to start containers.

> If the filesystem of the container is available on the local filesystem used by the plugin,

Reviewer: ```suggestion
If the filesystem of the container is available on the CernVM-FS repository used by the plugin,
```
